packages/zoltan2/test/helpers/AdapterForTests.hpp

```diff
@@ -120,10 +116,32 @@ class AdapterForTests{
    *
    * \return Ptr to the constructed adapter cast to the base class
    */
-  static AdapterWithOptionalCoordinateAdapter getAdapterForInput(
+  AdapterFactory(
     UserInputForTests *uinput, const ParameterList &pList,
     const RCP<const Comm<int> > &comm);
-  
+
+  ~AdapterFactory(); // handles deleting BaseAdapterRoot * data for adapter
+
+  AdapterWithOptionalCoordinateAdapter adaptersSet;
+
+  /*! \brief dynamic_cast adapter and then call getLocalNumIDs
+   */
+  size_t getLocalNumIDs() const {
+      #define GET_LOCAL_NUM_IDS(adapterClass)                                  \
+          return dynamic_cast<adapterClass*>(adaptersSet.main.adapter)->getLocalNumIDs();
+      Z2_TEST_UPCAST(adaptersSet.main.adapterType, GET_LOCAL_NUM_IDS);
+      throw std::logic_error( "getLocalNumIDs() failed to match adapter name" );
+  }
+
+  /*! \brief dynamic_cast adapter and then call getIDsView
+   */
+  void getIDsView(const zgno_t *&Ids) const {
```

Comment:
I'd like Michel to walk me through these routines.  I am not used to seeing macros inside functions like this; is it a typical factory pattern?  What are these functions doing?  How are they used?  Thanks.